pytorch_lightning/loops/epoch/prediction_epoch_loop.py

```diff
@@ -160,8 +162,11 @@ def _build_kwargs(self, batch: Any, batch_idx: int, dataloader_idx: int) -> Dict
 
     def _store_batch_indices(self, dataloader_idx: int) -> None:
         """Stores the batch indices if the predictions should be stored."""
-        batch_sampler = self.trainer.predict_dataloaders[dataloader_idx].batch_sampler
+        current_dataloader = self.trainer.predict_dataloaders[dataloader_idx]
+        batch_sampler = getattr(current_dataloader, "batch_sampler", None)
```

Comment:
```suggestion
        batch_sampler = current_dataloader.batch_sampler
```
the batch sampler is an attribute of Dataloader

Reply:
I thought so, but @ananthsub raised a concern about it.

Reply:
@ananthsub what was your concern? usually he is on the side of stricter typing ^^

Reply:
And it might not be true if a user doesn't provide a DataLoader class.

Reply:
custom dataloaders (including our internal one) might not subclass from torch.utils.data.dataloader, but only implement the iterator interface.

Reply:
@awaelchli , adding to @edward-io 's response, I think @tchaton 's fix here is the safest as it matches the current behavior around the batch sampler check here: https://github.com/PyTorchLightning/pytorch-lightning/blob/ec828b826717cd3b5beabcb6d0cacf41b2320a98/pytorch_lightning/trainer/data_loading.py#L160-L170

Reply:
ok but note pretty much everywhere in lightning we are expecting an instance of DataLoader. I would be highly surprised if you can run Lightning without that. We also have no tests for non-dataloader data iterables

Reply:
+1 to @awaelchli concerns... Pretty much all of our code in `data_loading.py` assumes a `DataLoader` instance is passed.

Properly supporting non-dataloaders should be part of a different larger discussion. This  `getattr` call here just makes this piece a bit more confusing for those reading through the loops.

Reply:
Done !